File: DailyChallenges/Scala/LeetCode_124.scala
object Solution {
    def maxPathSum(root: TreeNode): Int = {
        var ans = Int.MinValue
        def dfs(node: TreeNode): Int = if (node == null) 0 else {
            val l = math.max(0, dfs(node.left))
            val r = math.max(0, dfs(node.right))
            ans = math.max(ans, l + node.value + r);
            math.max(l, r) + node.value
        }
        dfs(root)
        ans
    }
}

File: DailyChallenges/CPP/LeetCode_124.cpp
// 124. Binary Tree Maximum Path Sum @ 2022/12/11
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        function<int(TreeNode*)> dfs = [&](TreeNode * node){
            if (!node) return 0;
            int l = dfs(node->left);
            int r = dfs(node->right);
            ans = max(ans, ((l>0)?l:0) + node->val + ((r>0)?r:0));
            return max(((l>0)?l:0), ((r>0)?r:0)) + node->val;
        };
        dfs(root);
        return ans;
    }
};
